### PR TITLE
C# discovery snippet generation

### PR DESCRIPTION
**Pull request by chrisdunelm**

### CODE REVIEW

File: src/main/java/com/google/api/codegen/csharp/CSharpContextCommon.java [hunk 1]
```diff
@@ -31,11 +31,7 @@
    */
   // TODO(jonskeet): Handle naming collisions.
   private final TreeSet<String> imports = new TreeSet<>();
-  private final String serviceNamespace;
-
-  public CSharpContextCommon(String serviceNamespace) {
-    this.serviceNamespace = serviceNamespace;
-  }
+  //private final String serviceNamespace;
```
Comment (jskeet) [inline src/main/java/com/google/api/codegen/csharp/CSharpContextCommon.java]: Delete commented-out field?

File: src/main/java/com/google/api/codegen/csharp/CSharpDiscoveryContext.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+  }
+
+  public String addImport(String namespace) {
+    return csharpCommon.addImport(namespace);
+  }
+  
+  private String packageNameAndImport() {
+    String packageName = CSharpContextCommon.s_underscoresToPascalCase(
+        getApiaryConfig().getServiceCanonicalName());
+    String versionName = getApiaryConfig().getServiceVersion().replace('.', '_');
```
Comment (jskeet) [inline src/main/java/com/google/api/codegen/csharp/CSharpDiscoveryContext.java]: I assume this is following the existing codegen? Would prefer to stick to just major version numbers in the namespace, but we're not changing the past. (Just checking that we're on the same page.)

File: src/main/java/com/google/api/codegen/csharp/CSharpSnippetSetRunner.java [hunk 3]
```diff
@@ -48,23 +49,20 @@ public GeneratedResult generate(
             snippetFileName,
             ImmutableMap.<String, Object>of("context", context));
 
-    // TODO don't depend on a cast here
-    CSharpGapicContext csharpContext = (CSharpGapicContext) context;
-
     String outputFilename = snippets.generateFilename(element).prettyPrint();
+    CSharpContextCommon csharpContextCommon = new CSharpContextCommon();
```
Comment (jskeet) [inline src/main/java/com/google/api/codegen/csharp/CSharpSnippetSetRunner.java]: I confess I have very little idea what these different contexts are about at this stage.

File: src/main/resources/com/google/api/codegen/csharp/wrapper.snip [hunk 4]
```diff
@@ -19,7 +19,7 @@
 
     // Generated code. DO NOT EDIT!
 
-    @join import : imports
+    @join import : context.removeItem(imports, context.getNamespace(service.getFile))
```
Comment (jskeet) [inline src/main/resources/com/google/api/codegen/csharp/wrapper.snip]: Why do we need to remove the import at all?

**Comment by jskeet** [comment]
Do we have any way of testing that generated samples all compile? If so, definitely LGTM. If not, have you tried some random sample of them manually?

File: src/main/java/com/google/api/codegen/csharp/CSharpContextCommon.java (hunk 1 above)
Comment (chrisdunelm) [inline src/main/java/com/google/api/codegen/csharp/CSharpContextCommon.java]: Done

File: src/main/java/com/google/api/codegen/csharp/CSharpDiscoveryContext.java (hunk 2 above)
Comment (chrisdunelm) [inline src/main/java/com/google/api/codegen/csharp/CSharpDiscoveryContext.java]: This is building the correct full name for the gRPC import, so cannot be changed,

File: src/main/java/com/google/api/codegen/csharp/CSharpSnippetSetRunner.java (hunk 3 above)
Comment (chrisdunelm) [inline src/main/java/com/google/api/codegen/csharp/CSharpSnippetSetRunner.java]: It's all been fairly significantly changed over the last couple of months. I've essentially just used the Java snippet code and changed it so it does what we need in C#

File: src/main/resources/com/google/api/codegen/csharp/wrapper.snip (hunk 4 above)
Comment (chrisdunelm) [inline src/main/resources/com/google/api/codegen/csharp/wrapper.snip]: It's to remove the "using" statement that imports the namespace within which this code is generated.
We used to do this in a different way, but the underlying framework has been altered so we can't do that any more :(

**Comment by chrisdunelm** [comment]
There is a framework to check all the generated snippets compile OK, but this PR doesn't integrate with it yet. Unfortunately it's a completely separate framework (written in go) that I don't yet have any understanding of :( I'll integrate with in as soon as I have time.
I will manually compile a few to check, then if all is OK commit this.

File: src/main/java/com/google/api/codegen/csharp/CSharpDiscoveryContext.java (hunk 2 above)
Comment (jskeet) [inline src/main/java/com/google/api/codegen/csharp/CSharpDiscoveryContext.java]: gRPC import? I thought this was for the REST API. The gRPC part *won't* include minor version numbers as far as I'm aware.

**Comment by chrisdunelm** [comment]
There were a couple of problems with imports. I've checked some random snippets compile, so if travis passes I'll merge this.

File: src/main/java/com/google/api/codegen/csharp/CSharpDiscoveryContext.java (hunk 2 above)
Comment (chrisdunelm) [inline src/main/java/com/google/api/codegen/csharp/CSharpDiscoveryContext.java]: Oops, yes, my mistake - it's the REST import.
As I understand it, most services do only include the major version.

Comment (chrisdunelm) [inline src/main/java/com/google/api/codegen/csharp/CSharpDiscoveryContext.java]: Yes, I've just checked, and most are major version only.
A counter-example is prediction, which is currently v1.6, but the namespace is Google.Apis.Prediction.v1_6. This code is to match that namespace when minor version is included.

File: src/main/java/com/google/api/codegen/ApiaryConfig.java [hunk 5]
```diff
@@ -141,6 +151,22 @@
     return mediaUpload;
   }
 
+  public String getServiceCanonicalName() {
+    return serviceCanonicalName;
+  }
+
+  public void setServiceCanonicalName(String serviceCanonicalName) {
+    this.serviceCanonicalName = serviceCanonicalName;
+  }
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/ApiaryConfig.java]: There weren't any set calls before your change here, why did you have to add them?

**Pull request merged**